docs/configuration.md
# Configuring the Ruby SDK

The Ruby SDK configuration block includes several different configuration options for:

1. OAuth 2.0 password grant
2. OAuth 2.0 authorization code grant
3. Configuration using `.env` file

In addition to OAuth 2.0, the SDK provides options for:

1. Logging
1. HTTP Retries
1. Loading via `.env` file

## Eample using OAuth 2.0 Password Grant

```ruby
require 'ringcentral_sdk'

client = RingCentralSdk::REST::Client.new do |config|
  # App info (mandatory)
  config.client_id = 'my_app_client_id'
  config.client_secret = 'my_app_client_secret'
  config.server_url = RingCentralSdk::RC_SERVER_SANDBOX

  # User info for password grant (optional)
  config.username = 'my_username'
  config.extension = 'my_extension'
  config.password = 'my_password'

  config.logger = Logger.new STDOUT
  config.retry = true
end
```

## Example using OAuth 2.0 Authorization Code Grant

```ruby
require 'ringcentral_sdk'

client = RingCentralSdk::REST::Client.new do |config|
  # App info (mandatory)
  config.client_id = 'my_app_client_id'
  config.client_secret = 'my_app_client_secret'
  config.server_url = RingCentralSdk::RC_SERVER_SANDBOX

  # OAuth 2.0 authorization code grant (optional)
  # Can be set later
  config.redirect_url = 'http://example.com/oauth'

  config.logger = Logger.new STDOUT
  config.retry = true
  config.retry_opts = { retry_after: 15, error_codes: [429, 503, 504] }
end
```

## Logging

The RingCentral SDK either takes a Logger object or will create a default `Logger` object. The same object is passed to dependencies including the faraday HTTP client and the faraday request retry middleware gem. This is set using the logger config variable:

```ruby
client = RingCentral::REST::Client.new do |config|
  config.logger = Logger.new STDOUT
  ...
end
```

## HTTP Request Retries 

RingCentral SDK uses `FaradayMiddleware::Request::Retry` to handle HTTP request retries. By default it will retry on 429, 503 and 504 responses. A `Retry-After` header in seconds will be respected, falling back to a default retry after. The following can be set:

* `:error_codes` to retry, default `[429, 503, 504]`
* `:retry_after` seconds, default `10`

Note, there may be issues replay certain types of requests, notably `Faraday::UploadIO` requests. To reply these requests you can use `FaradayMiddleware::Request::RetryUtil` separately as seen in [RingCentral::Avatars::Creator](https://github.com/grokify/ringcentral-avatars-ruby/blob/master/lib/ringcentral-avatars/creator.rb).

## Using Environment Variables

When deploying an app, it can be useful to store some variables outside of the app code. This is especially useful for app and user credentials that should not be checked into source control systems. The RingCentral SDK supports reading some information from the environment, either directly or via a `.env` file. To use this capability set `config.load_env = true`.

The following environment variables will be loaded:

| ENV variable | Config setting | Notes |
|--------------|----------------|-------|
| `RINGCENTRAL_SERVER_URL` | `config.server_url` | Must be set to value of URL, not constant |
| `RINGCENTRAL_CLIENT_ID` | `config.client_id` |
| `RINGCENTRAL_CLIENT_SECRET` | `config.client_secret` |
| `RINGCENTRAL_REDIRECT_URL` | `config.app_redirect_url` |
| `RINGCENTRAL_USERNAME` | `config.username` |
| `RINGCENTRAL_EXTENSION` | `config.extension` |
| `RINGCENTRAL_PASSWORD` | `config.password` |
| `RINGCENTRAL_TOKEN` | `config.token` | JSON token response |
| `RINGCENTRAL_TOKEN_FILE` | `config.token_file` | File containing JSON token response |
| `RINGCENTRAL_RETRY` | `config.retry` | Set to `true` or `false` |
| `RINGCENTRAL_RETRY_OPTIONS` | `config.retry_options | JSON string for `FaradayMiddleware::Request::Retry` options |

Ruby file:

```ruby
require 'ringcentral_sdk'

client = RingCentralSdk::REST::Client.new do |config|
  config.load_env = true
end
```

Example Environment Variables / .env file

```
RINGCENTRAL_SERVER_URL=https://platform.ringcentral.com
RINGCENTRAL_CLIENT_ID=my_app_client_id
RINGCENTRAL_CLIENT_SECRET=my_app_client_secret
RINGCENTRAL_USERNAME=my_username
RINGCENTRAL_EXTENSION=my_extension
RINGCENTRAL_PASSWORD=my_password
```


scripts/rc_config_sample.env.txt
RC_APP_KEY=myAppKey
RC_APP_SECRET=myAppSecret
RC_APP_SERVER_URL=myRcApiServerUrl
RC_APP_REDIRECT_URL=myRcApiRedirectUrl

RC_USER_USERNAME=myUsername
RC_USER_EXTENSION=myExtension
RC_USER_PASSWORD=myPassword

RC_OAP_SERVER_URL=myOnlineAccountPortalServerUrl

RC_DEMO_FAX_TO=demoFaxToPhoneNumber
RC_DEMO_FAX_COVERPAGE_TEXT=demoFaxCoverPageText
RC_DEMO_FAX_FILE=demoFaxFilepath

RC_DEMO_SMS_TO=+16505551212
RC_DEMO_SMS_FROM=+14155551212
RC_DEMO_SMS_TEXT=Test SMS from Ruby!
RC_DEMO_MMS_FILE=glip_chatbots_700x400.png

RC_DEMO_VB_API_KEY=myVoiceBaseApiKey
RC_DEMO_VB_PASSWORD=myVoiceBasePassword

RC_DEMO_GLIP_WEBHOOK_URL=https://hooks.glip.com/webhook/11111111-2222-3333-4444-555555555555
RC_DEMO_GLIP_WEBHOOK_ICON=http://example.com/icon.png


CHANGELOG.md
CHANGELOG
---------
- **2017-06-20**: 2.2.1
  - Additions
    - Generic Multipart helper
  - Bugfixes
    - Fix `client.authorize_url`
    - Fix OAuth2 Sintra demo
- **2017-04-01**: 2.1.0
  - Additions
    - Add SMS/MMS helper and demo
  - Update
    - Refactor Fax helper
- **2016-12-28**: 2.0.2
  - Update
    - Update to add retry for `504` error code as default in addition to `429`, `503`
- **2016-12-28**: 2.0.1
  - Minor fixes
- **2016-12-28**: 2.0.0
  - Breaking Changes
    - New block-based config, replacing old method param and env config
  - Additions
    - Add auto-retry request using `FaradayMiddleware::Request::Retry`
    - Add configurable logger
  - Update
    - Update `pubnub` dependency to `4.0.16` from `3.8.x`
    - Update `rake` dependency to `12.x.x` from `11.x.x`
    - Update gemspec to include dependency versions
  - Bugfixes
    - Fix `logger.level` default setting
- **2016-09-05**: 1.3.4
  - Bugfixes
    - Fix `RingCentralSdk::REST::Subscription` renewal bug
- **2016-09-04**: 1.3.2 & 1.3.3
  - Bugfixes
    - Fix `RingCentralSdk::REST::Subscription` renewal bug
- **2016-08-24**: 1.3.1
  - Updates
    - Lock PubNub dependency to 3.x due to back compat issue with 4.x
- **2016-08-23**: 1.3.0
  - Additions
    - Add OAuth 2.0 parameter support for `authorize_(code|password|user)` methods
  - Bugfixes
    - Fix `Client#token` to return token used by `FaradayMiddleware::OAuth2Refresh`
- **2016-06-01**: 1.2.3
  - Updates
    - Update subscription for new header requirement
- **2016-05-30**: 1.2.2
  - Additions
    - Add `RingCentralSdk::REST::Request::Fax` to support all `MIME::Media` MIME parts
    - Add docs for call queue add / remove agents
    - Add docs for fax `multipart/form-data` support
    - Add demo for fax custom cover page
- **2016-04-23**: 1.2.1
  - Updates
    - Update `RingCentral::REST::Client` `app_config =` to `set_app_config`
- **2016-04-23**: 1.2.0
  - Additions
    - Add `RingCentral::REST::Client` custom instance headers support
    - Add `RingCentral::REST::Client` `attr_reader` for `app_config`
    - Add docs for Answering Rules
  - Updates
    - Update Pubnub dependency to 3.8.0 from 3.7.3
  - Removals
    - Deprecate Ruby 1.9.3 due to `mime-types-data` (3.2016.0221) requiring Ruby >= 2.0
- **2016-02-13**: 1.1.1
  - Additions
    - Add `client.send_request` hash support in addition to `Request::Base` object
    - Add documentation / FAQ for fax `to.name` attribute
  - Bugfixes
    - Fix `RingCentralSdk::REST::Cache::Extensions` bug - old client
    - Fix `RingCentralSdk::REST::Request::Simple` bug - initialize body
- **2016-02-07**: 1.1.0
  - Add Subscription to Retrieve New Messages functionality
    - `./lib`
      - Add `RingCentralSdk::REST::Event`
      - Add `RingCentralSdk::REST::MessagesRetriever`
      - Add `RingCentralSdk::REST::Request::Simple`
      - Update `RingCentralSdk::REST::Client` - update `send_request`
      - Update `RingCentralSdk::REST::Request::Base` - add `headers`
    - `./scripts`
      - Add `scripts/fax_download.rb` demo script
      - Add `scripts/fax_download_new.rb` demo script
      - Add `scripts/sms_to_chat.rb` demo script to send inbound SMS to chat
- **2016-01-26**: 1.0.0
  - Major refactor focusing on naming changes to improve consistency
    - All code moved into `RingCentralSdk::REST` namespace
    - Update: rename `RingCentralSdk::Platform` to `RingCentralSdk::REST::Client`
    - Update: rename `RingCentralSdk::Platform::client` to `RingCentralSdk::REST::Client::http`
    - Update: rename `RingCentralSdk::Helpers` namespace to `RingCentralSdk::Request`
  - Add `RingCentralSdk::REST::Config` to support `dotenv` config files
    - Add `RingCentralSdk::REST::Config` reads dotenv files
    - Add `scripts/account-extensions.rb` script
    - Update sample scripts to use new Config object
  - Add `RingCentral::REST::Messages`, `RingCentral::REST::Messages::MessagesSMS` and `RingCentral::REST::Messages::MessagesFax`
  - Update `RingCentralSdk::REST::Request::Fax` helper for nicer interface, formerly `RingCentralSdk::Helpers::Fax`
- **2016-01-22**: 0.5.2
  - Add Call Queue Management support
    - Add `RingCentralSdk::Cache::Extensions`
    - Add `RingCentralSdk::Helpers::ExtensionPresence`
    - Add Call Queue Member Status mydocs documentation
- **2015-12-08**: 0.5.1
  - Add call recording transcription demo script via VoiceBase
  - Fix PubNub unsubscribe bug
  - Fix fax helper to support UTF-8 metadata with file attachment
  - Update tests to support access token without refresh token
  - Update sample scripts credentials file format to be more scalable
- **2015-10-18**: 0.5.0
  - Add stubbed tests via `mocha`
  - Update naming and cleanup interface
    - Add `RingCentralSdk.new()` sugar method
    - Update `RingCentralSdk::RC_SERVER_PRODUCTION` url from `RingCentralSdk::Sdk::RC_SERVER_Production`
    - Delete `lib/ringcentral_sdk/sdk.rb` file and `RingCentralSdk::Sdk` class
    - Delete `lib/ringcentral_sdk/version.rb` file
    - Delete `VERSION.txt` file
- **2015-10-15**: 0.4.0
  - Add OAuth `authorization_code` support
    - `./lib`
      - Add 3-Legged OAuth support via platform `authorize_url` and `authorize_code` methods
      - Update SDK initialization by converting optional params to options hash
      - Update `RingCentralSdk::Platform::Platform` to `RingCentralSdk::Platform`
    - `./scripts`
      - Add `oauth2-sinatra` demo app in `scripts` directory
      - Add `call_recording_download.rb` demo script
      - Add `rc-credentials.json` separate demo credentials file
      - Update `fax_send.rb` demo script with name change
- **2015-10-07**: 0.3.0
  - Add subscription support via PubNub
  - Add `ruby-head`, `jruby`, `jruby-head`, and `rbx-2` to Travis CI
  - Remove Ruby 1.8.7 support due to `hitimes` requirement in `pubnub` and `eventmachine`
- **2015-09-14**: 0.2.0
  - Refactor platform class to remove `authorized` method in favor of `set_token`
  - Add `User-Agent` header support

lib/ringcentral_sdk/rest/configuration.rb
require 'dotenv'
require 'fiddle'
require 'logger'
require 'multi_json'
require 'ringcentral_sdk'
require 'uri'

module RingCentralSdk
  module REST
    # Configuration class populated by Client constructor block
    class Configuration
      attr_accessor :server_url
      attr_accessor :client_id
      attr_accessor :client_secret
      attr_accessor :redirect_url

      attr_accessor :username
      attr_accessor :extension
      attr_accessor :password
      attr_accessor :token
      attr_accessor :token_file

      attr_accessor :load_env
      attr_accessor :headers
      attr_accessor :retry
      attr_accessor :retry_options
      attr_accessor :logger

      def inflate
        @logger = default_logger if !defined?(@logger) || @logger.nil?
        load_environment if load_env
        inflate_headers
        inflate_retry
        inflate_retry_options
        inflate_token
      end

      def default_logger
        logger = Logger.new STDOUT
        logger.level = Logger::WARN
        logger
      end

      def load_environment
        Dotenv.load
        @server_url    = ENV['RINGCENTRAL_SERVER_URL']    if ENV.key? 'RINGCENTRAL_SERVER_URL'
        @client_id     = ENV['RINGCENTRAL_CLIENT_ID']     if ENV.key? 'RINGCENTRAL_CLIENT_ID'
        @client_secret = ENV['RINGCENTRAL_CLIENT_SECRET'] if ENV.key? 'RINGCENTRAL_CLIENT_SECRET'
        @redirect_url  = ENV['RINGCENTRAL_REDIRECT_URL']  if ENV.key? 'RINGCENTRAL_REDIRECT_URL'
        @username      = ENV['RINGCENTRAL_USERNAME']      if ENV.key? 'RINGCENTRAL_USERNAME'
        @extension     = ENV['RINGCENTRAL_EXTENSION']     if ENV.key? 'RINGCENTRAL_EXTENSION'
        @password      = ENV['RINGCENTRAL_PASSWORD']      if ENV.key? 'RINGCENTRAL_PASSWORD'
        @token         = ENV['RINGCENTRAL_TOKEN']         if ENV.key? 'RINGCENTRAL_TOKEN'
        @token_file    = ENV['RINGCENTRAL_TOKEN_FILE']    if ENV.key? 'RINGCENTRAL_TOKEN_FILE'
        @retry         = ENV['RINGCENTRAL_RETRY']         if ENV.key? 'RINGCENTRAL_RETRY'
        @retry_options = ENV['RINGCENTRAL_RETRY_OPTIONS'] if ENV.key? 'RINGCENTRAL_RETRY_OPTIONS'
        @headers       = ENV['RINGCENTRAL_HEADERS']       if ENV.key? 'RINGCENTRAL_HEADERS'
      end

      def inflate_retry
        if !defined?(@retry) || @retry.nil?
          @retry = false
        elsif @retry.is_a? String
          @retry = @retry.to_s.strip.downcase == 'true' ? true : false
        elsif ![true, false].include? @retry
          @retry = @retry ? true : false
        end
      end

      def inflate_retry_options
        if @retry == false
          @retry_options = {}
          return
        end
        if !@retry_options.nil? && @retry_options.to_s =~ /^\s*{/
          @retry_options = MultiJson.decode @retry_options.to_s, symbolize_keys: true
        else
          @retry_options = {}
        end
        @retry_options[:error_codes] = [429, 503, 504] unless @retry_options.key? :error_codes
        @retry_options[:logger] = @logger
      end

      def inflate_headers
        @headers = {} unless defined? @headers
        if !@headers.nil? && @headers.is_a?(String) && @headers =~ /^\s*{/
          @headers = MultiJson.decode @headers, symbolize_keys: true
        end
      end

      def inflate_token
        @token = nil unless defined? @token

        if (@token.nil? || @token.empty?) && !token_file.nil? && !@token_file.empty?
          @token = IO.read @token_file if File.exist? @token_file
        end

        if !defined?(@token) && !@token.nil? && @token.is_a?(String) && @token =~ /^\s*{/
          @token = MultiJson.decode @token
        end
      end

      def authorize_url
        URI.join @server_url, RingCentralSdk::REST::Client::AUTHZ_ENDPOINT
      end
    end
  end
end
